test/e2e_env/universal/gateway/gateway.go

```diff
@@ -0,0 +1,267 @@
+)
+
+func Gateway() {
+	var mesh = "gateway"
+	var mtlsMesh = "gateway-mtls"
+
+	const gatewayPort = 8080
+	const mtlsGatewayPort = 8081
+
+	ExternalServerUniversal := func(mesh string) InstallFunc {
+		return func(cluster Cluster) error {
+			return cluster.DeployApp(
+				WithArgs([]string{"test-server", "echo", "--port", "8080", "--instance", "external-echo"}),
+				WithName("external-echo-"+mesh),
+				WithMesh(mesh),
+				WithoutDataplane(),
+				WithVerbose())
+		}
+	}
+
+	// DeployCluster creates a universal Kuma cluster using the
+	// provided options, installing an echo service as well as a
+	// gateway and a client container to send HTTP requests.
+	BeforeAll(func() {
+		setup := NewClusterSetup().
+			Install(MeshUniversal(mesh)).
+			Install(MTLSMeshUniversal(mtlsMesh)).
+			Install(gatewayClientAppUniversal("gateway-client")).
+			Install(echoServerApp(mesh, "echo-server", "echo-service", "universal")).
+			Install(echoServerApp(mtlsMesh, "echo-server-mtls", "echo-service", "universal")).
+			Install(GatewayProxyUniversal(mesh, "gateway-proxy")).
+			Install(GatewayProxyUniversal(mtlsMesh, "gateway-proxy-mtls")).
+			Install(ExternalServerUniversal(mesh))
+
+		Expect(setup.Setup(env.Cluster)).To(Succeed())
+
+		Expect(
+			env.Cluster.GetKumactlOptions().KumactlApplyFromString(MkGateway("edge-gateway", mesh, false, "example.kuma.io", "echo-service", gatewayPort)),
+		).To(Succeed())
+
+		Expect(
+			env.Cluster.GetKumactlOptions().KumactlApplyFromString(MkGateway("edge-gateway", mtlsMesh, false, "example-mtls.kuma.io", "echo-service", mtlsGatewayPort)),
+		).To(Succeed())
+	})
+
+	E2EAfterAll(func() {
+		Expect(env.Cluster.DeleteMeshApps(mesh)).To(Succeed())
+		Expect(env.Cluster.DeleteMesh(mesh)).To(Succeed())
+		Expect(env.Cluster.DeleteMeshApps(mtlsMesh)).To(Succeed())
+		Expect(env.Cluster.DeleteMesh(mtlsMesh)).To(Succeed())
+	})
+
+	GatewayAddress := func(appName string) string {
+		return env.Cluster.GetApp(appName).GetIP()
+	}
+
+	GatewayAddressPort := func(appName string, port int) string {
+		ip := env.Cluster.GetApp(appName).GetIP()
+		return net.JoinHostPort(ip, strconv.Itoa(port))
+	}
+
+	Context("when mTLS is disabled", func() {
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "universal",
+				GatewayAddressPort("gateway-proxy", gatewayPort), "example.kuma.io")
+		})
+	})
+
+	Context("when mTLS is enabled", func() {
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "universal",
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+
+		// In mTLS mode, only the presence of TrafficPermission rules allow services to receive
+		// traffic, so removing the permission should cause requests to fail. We use this to
+		// prove that mTLS is enabled
+		It("should fail without TrafficPermission", func() {
+			proxyRequestsWithMissingPermission(env.Cluster, mtlsMesh,
```

Comment:
I think this helper is overkill. It's used once and the test logic is basically squeezed there
```
// when traffic permission is removed
// then traffic is blocked
```

can we inline it?